[BUG]: STM reading inconsistent values

### Version

0.18.0

### Scala Version

3.6.4

### Expected Behavior

Since the 2 `TRef`s initially contain equal values, and I only ever write into them equal values, I expect never to observe non-equal values in a single transaction. That is, I expect the program to exit without an exception.

### Actual Behavior

The program exits with an exception like this:
```
Exception in thread "main" java.lang.AssertionError: a111111111111111111111111 != a1111111111111111111111111
```

### Steps to Reproduce

```scala
object StmTest extends KyoApp {
  run {
    for {
      // initially they contain equal values:
      r1 <-STM.run(TRef.init("a"))
      r2 <-STM.run(TRef.init("a"))
      txn1 = for {
        v1 <- r1.get
        v2 <- r2.get
        // we only write equal values:
        _ <- r1.set(v1 + "1")
        _ <- r2.set(v2 + "1")
      } yield ()
      txn2 = r1.get.map { v1 =>
        r2.get.map { v2 =>
          if (v1 == v2) {
            ()
          } else {
            // observed non-equal values:
            Abort.panic(new AssertionError(s"${v1} != ${v2}")) // <- fails here
          }
        }
      }
      once = Async.zip(STM.run(txn1), STM.run(txn2))
      _ <- Async.foreachDiscard(1 to 100000, concurrency = 8) { _ => once }
    } yield ()
  }
}
```

### Current Workaround

_No response_

--- a/kyo-prelude/shared/src/main/scala/kyo/Abort.scala
+++ b/kyo-prelude/shared/src/main/scala/kyo/Abort.scala
@@ -361,6 +361,27 @@ object Abort:
             case Failure(e) => onFail(e)
             case Panic(thr) => throw thr
 
+    /** Recovers from an Abort failure or panic by applying the provided function to the Error value.
+      *
+      * This method allows you to handle both failures and panics in an Abort effect with a single handler function that receives the
+      * complete Error value (either Failure or Panic).
+      *
+      * @param onError
+      *   A function that takes the Error value (either Failure[E] or Panic) and returns a new computation
+      * @param v
+      *   The original computation that may fail or panic
+      * @return
+      *   A computation that either succeeds with the original value or the recovered value
+      */
+    def recoverError[E](
+        using Frame
+    )[A, B, S, ER](onError: Error[E] => B < S)(v: => A < (Abort[E | ER] & S))(
+        using
+        ct: SafeClassTag[E],
+        reduce: Reducible[Abort[ER]]
+    ): (A | B) < (S & reduce.SReduced & Abort[Nothing]) =
+        runWith[E](v)(_.foldError(identity, onError))
+
     /** Recovers from an Abort failure by applying the provided function.
       *
       * This method allows you to handle failures in an Abort effect and potentially continue the computation with a new value. It only
@@ -440,6 +461,32 @@ object Abort:
             case Success(a) => onSuccess(a)
             case Failure(e) => onFail(e)
             case Panic(thr) => throw thr
+
+    /** Handles an Abort effect by applying the provided functions to the success value or error.
+      *
+      * This method allows you to handle both the success case and error cases (both failures and panics) in an Abort effect. Unlike
+      * [[fold]], this method provides the complete Error value to the error handler function.
+      *
+      * @param onSuccess
+      *   A function that takes the success value of type A and returns a new computation
+      * @param onError
+      *   A function that takes the Error value (either Failure[E] or Panic) and returns a new computation
+      * @param v
+      *   The original computation that may fail
+      * @return
+      *   A computation that applies the appropriate handler function based on the result
+      */
+    def foldError[E](
+        using Frame
+    )[A, B, S, ER](
+        onSuccess: A => B < S,
+        onError: Error[E] => B < S
+    )(v: => A < (Abort[E | ER] & S))(
+        using
+        ct: SafeClassTag[E],
+        reduce: Reducible[Abort[ER]]
+    ): B < (S & reduce.SReduced) =
+        runWith[E](v)(_.foldError(onSuccess, onError))
 
     /** Catches exceptions of type E and converts them to Abort failures.
       *

--- a/kyo-prelude/shared/src/test/scala/kyo/AbortTest.scala
+++ b/kyo-prelude/shared/src/test/scala/kyo/AbortTest.scala
@@ -1330,4 +1330,87 @@ class AbortTest extends Test:
         }
     }
 
+    "recoverError" - {
+        case class CustomError(message: String) derives CanEqual
+
+        "handles failures" in {
+            val computation = Abort.fail(CustomError("Expected error"))
+            val recovered =
+                Abort.recoverError[CustomError] {
+                    error => s"Recovered: ${error.show}"
+                }(computation)
+
+            assert(Abort.run(recovered).eval == Result.succeed("Recovered: Failure(CustomError(Expected error))"))
+        }
+
+        "handles panics" in {
+            val ex          = new RuntimeException("Panic message")
+            val computation = Abort.panic(ex)
+            val recovered = Abort.recoverError[CustomError] {
+                error => s"Recovered: ${error.show}"
+            }(computation)
+
+            assert(Abort.run(recovered).eval == Result.succeed("Recovered: Panic(java.lang.RuntimeException: Panic message)"))
+        }
+
+        "doesn't affect successful computations" in {
+            val computation: String < Abort[CustomError] = "success"
+            var called                                   = false
+            val recovered =
+                Abort.recoverError[CustomError] { _ =>
+                    called = true
+                    "Should not be called"
+                }(computation)
+
+            assert(called == false)
+            assert(Abort.run(recovered).eval == Result.succeed("success"))
+        }
+    }
+
+    "foldError" - {
+        case class CustomError(message: String) derives CanEqual
+
+        "handles success case" in {
+            val computation: Int < Abort[CustomError] = 42
+            val result = Abort.foldError[CustomError](
+                onSuccess = i => s"Success: $i",
+                onError = error => s"Error: ${error.show}"
+            )(computation)
+
+            assert(result.eval == "Success: 42")
+        }
+
+        "handles failure case" in {
+            val computation = Abort.fail(CustomError("Expected error"))
+            val result = Abort.foldError[CustomError](
+                onSuccess = i => s"Success: $i",
+                onError = error => s"Error: ${error.show}"
+            )(computation)
+
+            assert(result.eval == "Error: Failure(CustomError(Expected error))")
+        }
+
+        "handles panic case" in {
+            val ex          = new RuntimeException("Panic message")
+            val computation = Abort.panic(ex)
+            val result = Abort.foldError[CustomError](
+                onSuccess = i => s"Success: $i",
+                onError = error => s"Error: ${error.show}"
+            )(computation)
+
+            assert(result.eval == "Error: Panic(java.lang.RuntimeException: Panic message)")
+        }
+
+        "removes Abort from the effect set" in {
+            val computation = Abort.fail(CustomError("Expected error"))
+            val folded = Abort.foldError[CustomError](
+                onSuccess = i => s"Success: $i",
+                onError = _ => "Error handled"
+            )(computation)
+
+            val _: String < Any = folded
+            succeed
+        }
+    }
+
 end AbortTest

--- a/kyo-stm/shared/src/main/scala/kyo/STM.scala
+++ b/kyo-stm/shared/src/main/scala/kyo/STM.scala
@@ -251,12 +251,12 @@ end STM
 
 /** A FailedTransaction exception that is thrown when a transaction fails to commit. Contains the frame where the failure occurred.
   */
-class FailedTransaction(error: Maybe[Result.Error[?]] = Absent)(using Frame)
+final class FailedTransaction(error: Maybe[Result.Error[?]] = Absent)(using Frame)
     extends KyoException(
         s"STM transaction failed!",
         error.fold("") {
             _.failureOrPanic match
                 case ex: Throwable => ex
-                case _             => s"Error: ${error.show}"
+                case _             => error.show
         }
     )

--- a/kyo-stm/shared/src/main/scala/kyo/STM.scala
+++ b/kyo-stm/shared/src/main/scala/kyo/STM.scala
@@ -163,9 +163,9 @@ object STM:
                     var i = 0
                     // Pre-validate and dump the log to the flat array
                     logMap.foreachEntry { (ref, entry) =>
-                        // This code uses exception throwing because
+                        // This code uses `boundary`/`break` because
                         // foreachEntry is the only way to traverse the
-                        // map without allocating tuples, so throwing
+                        // map without allocating tuples, so throwing via `break`
                         // is the workaround to short circuit
                         if !ref.validate(entry) then boundary.break(false)
                         array(i) = ref

--- a/kyo-stm/shared/src/main/scala/kyo/STM.scala
+++ b/kyo-stm/shared/src/main/scala/kyo/STM.scala
@@ -2,10 +2,7 @@ package kyo
 
 import kyo.Result.Failure
 import scala.annotation.tailrec
-
-/** A FailedTransaction exception that is thrown when a transaction fails to commit. Contains the frame where the failure occurred.
-  */
-class FailedTransaction()(using Frame) extends KyoException
+import scala.util.boundary
 
 /** Software Transactional Memory (STM) provides concurrent access to shared state using optimistic locking. Rather than acquiring locks
   * upfront, transactions execute speculatively and automatically retry if conflicts are detected during commit. While this enables better
@@ -70,7 +67,7 @@ object STM:
       * @return
       *   The result of the computation if successful
       */
-    def run[E, A, S](
+    def run[E: SafeClassTag, A, S](
         using Isolate.Stateful[S, Async & Abort[E | FailedTransaction]]
     )(v: A < (STM & Abort[E] & Async & S))(using frame: Frame): A < (S & Async & Abort[E | FailedTransaction]) =
         run(defaultRetrySchedule)(v)
@@ -84,7 +81,7 @@ object STM:
       * @return
       *   The result of the computation if successful
       */
-    def run[E, A, S](
+    def run[E: SafeClassTag, A, S](
         using isolate: Isolate.Stateful[S, Async & Abort[E | FailedTransaction]]
     )(retrySchedule: Schedule)(v: A < (STM & Abort[E] & Async & S))(
         using frame: Frame
@@ -93,100 +90,38 @@ object STM:
             isolate.restore(run(retrySchedule)(isolate.isolate(st, v)))
         }
 
-    private def run[E, A](retrySchedule: Schedule)(v: A < (STM & Abort[E] & Async))(
+    private def run[E: SafeClassTag, A](retrySchedule: Schedule)(v: A < (STM & Abort[E] & Async))(
         using Frame
     ): A < (Async & Abort[E | FailedTransaction]) =
         TID.useIO {
             case -1L =>
-                // New transaction without a parent, use regular commit flow
-                Retry[FailedTransaction](retrySchedule) {
-                    TID.useNew { tid =>
-                        Var.runWith(TRefLog.empty)(v) { (log, result) =>
-                            val logMap = log.toMap
-                            logMap.size match
-                                case 0 =>
-                                    // Nothing to commit
-                                    result
-                                case 1 =>
-                                    // Fast-path for a single ref
-                                    IO.Unsafe {
-                                        val (ref, entry) = logMap.head
-                                        // No need to pre-validate since `lock` validates and
-                                        // there's a single ref
-                                        if ref.lock(entry) then
-                                            ref.commit(tid, entry)
-                                            ref.unlock(entry)
-                                            result
-                                        else
-                                            Abort.fail(FailedTransaction())
-                                        end if
-                                    }
-                                case size =>
-                                    // Commit multiple refs
-                                    IO.Unsafe {
-                                        // Flattened representation of the log
-                                        val array = new Array[Any](size * 2)
-
-                                        try
-                                            def fail = throw new FailedTransaction()
-
-                                            var i = 0
-                                            // Pre-validate and dump the log to the flat array
-                                            logMap.foreachEntry { (ref, entry) =>
-                                                // This code uses exception throwing because
-                                                // foreachEntry is the only way to traverse the
-                                                // map without allocating tuples, so throwing
-                                                // is the workaround to short circuit
-                                                if !ref.validate(entry) then fail
-                                                array(i) = ref
-                                                array(i + 1) = entry
-                                                i += 2
-                                            }
-
-                                            // Sort references by identity to prevent deadlocks
-                                            quickSort(array, size)
-
-                                            // Convenience accessors to the flat log
-                                            inline def ref(idx: Int)   = array(idx * 2).asInstanceOf[TRef[Any]]
-                                            inline def entry(idx: Int) = array(idx * 2 + 1).asInstanceOf[TRefLog.Entry[Any]]
-
-                                            @tailrec def lock(idx: Int): Int =
-                                                if idx == size then size
-                                                else if !ref(idx).lock(entry(idx)) then idx
-                                                else lock(idx + 1)
-
-                                            @tailrec def unlock(idx: Int, upTo: Int): Unit =
-                                                if idx < upTo then
-                                                    ref(idx).unlock(entry(idx))
-                                                    unlock(idx + 1, upTo)
-
-                                            @tailrec def commit(idx: Int): Unit =
-                                                if idx < size then
-                                                    ref(idx).commit(tid, entry(idx))
-                                                    commit(idx + 1)
-
-                                            val acquired = lock(0)
-                                            if acquired != size then
-                                                // Failed to acquire some locks - rollback and retry
-                                                unlock(0, acquired)
-                                                fail
-                                            end if
-
-                                            // Successfully locked all references - commit changes
-                                            commit(0)
-
-                                            // Release all locks
-                                            unlock(0, size)
-                                            result
-                                        catch
-                                            case ex: FailedTransaction =>
-                                                Abort.fail(ex)
-                                        end try
-                                    }
-                            end match
+                TID.useNew { tid =>
+                    v.handle(
+                        Abort.recoverError[E] { error =>
+                            // Retry arbitrary E failures in case the transaction is inconsistent
+                            Var.use[TRefLog] { log =>
+                                IO.Unsafe {
+                                    if !commit(tid, log, probe = true) then
+                                        // The ref log shows inconsistency, retry the transaction
+                                        Abort.fail(FailedTransaction(Present(error)))
+                                    else
+                                        // No inconsistency detected, just propagate the error
+                                        Abort.error(error)
+                                }
+                            }
+                        },
+                        Var.runTuple(TRefLog.empty)
+                    ).map { (log, result) =>
+                        IO.Unsafe {
+                            if !commit(tid, log) then
+                                Abort.fail(FailedTransaction())
+                            else
+                                result
                         }
                     }
-                }
+                }.handle(
+                    Retry[FailedTransaction](retrySchedule)
+                )
             case parent =>
                 // Nested transaction inherits parent's transaction context but isolates RefLog.
                 // On success: changes propagate to parent. On failure: changes are rolled back
@@ -201,6 +136,81 @@ object STM:
         }
 
     end run
+
+    private def commit[A, S](tid: Long, log: TRefLog, probe: Boolean = false)(using AllowUnsafe): Boolean =
+        val logMap = log.toMap
+        logMap.size match
+            case 0 =>
+                // Nothing to commit
+                true
+            case 1 =>
+                // Fast-path for a single ref
+                val (ref, entry) = logMap.head
+                // No need to pre-validate since `lock` validates and
+                // there's a single ref
+                val ok = ref.lock(entry)
+                if ok then
+                    if !probe then ref.commit(tid, entry)
+                    ref.unlock(entry)
+                ok
+            case size =>
+                // Commit multiple refs
+                // Flattened representation of the log
+                val array = new Array[Any](size * 2)
+
+                boundary {
+
+                    var i = 0
+                    // Pre-validate and dump the log to the flat array
+                    logMap.foreachEntry { (ref, entry) =>
+                        // This code uses exception throwing because
+                        // foreachEntry is the only way to traverse the
+                        // map without allocating tuples, so throwing
+                        // is the workaround to short circuit
+                        if !ref.validate(entry) then boundary.break(false)
+                        array(i) = ref
+                        array(i + 1) = entry
+                        i += 2
+                    }
+
+                    // Sort references by identity to prevent deadlocks
+                    quickSort(array, size)
+
+                    // Convenience accessors to the flat log
+                    inline def ref(idx: Int)   = array(idx * 2).asInstanceOf[TRef[Any]]
+                    inline def entry(idx: Int) = array(idx * 2 + 1).asInstanceOf[TRefLog.Entry[Any]]
+
+                    @tailrec def lock(idx: Int): Int =
+                        if idx == size then size
+                        else if !ref(idx).lock(entry(idx)) then idx
+                        else lock(idx + 1)
+
+                    @tailrec def unlock(idx: Int, upTo: Int): Unit =
+                        if idx < upTo then
+                            ref(idx).unlock(entry(idx))
+                            unlock(idx + 1, upTo)
+
+                    @tailrec def commit(idx: Int): Unit =
+                        if idx < size then
+                            ref(idx).commit(tid, entry(idx))
+                            commit(idx + 1)
+
+                    val acquired = lock(0)
+                    if acquired != size then
+                        // Failed to acquire some locks - rollback and retry
+                        unlock(0, acquired)
+                        boundary.break(false)
+                    end if
+
+                    // Successfully locked all references - commit changes
+                    if !probe then commit(0)
+
+                    // Release all locks
+                    unlock(0, size)
+                    true
+                }
+        end match
+    end commit
 
     private def quickSort(array: Array[Any], size: Int): Unit =
         def swap(i: Int, j: Int): Unit =
@@ -238,3 +248,15 @@ object STM:
             loop(0, size - 1)
     end quickSort
 end STM
+
+/** A FailedTransaction exception that is thrown when a transaction fails to commit. Contains the frame where the failure occurred.
+  */
+class FailedTransaction(error: Maybe[Result.Error[?]] = Absent)(using Frame)
+    extends KyoException(
+        s"STM transaction failed!",
+        error.fold("") {
+            _.failureOrPanic match
+                case ex: Throwable => ex
+                case _             => s"Error: ${error.show}"
+        }
+    )

--- a/kyo-stm/shared/src/main/scala/kyo/TRef.scala
+++ b/kyo-stm/shared/src/main/scala/kyo/TRef.scala
@@ -148,6 +148,14 @@ final private class TRefImpl[A] private[kyo] (initialState: Write[A])
                 super.set(0)
         end match
     end unlock
+
+    override def toString() =
+        val lockState = get() match
+            case 0            => "free"
+            case Int.MaxValue => "writer"
+            case i            => s"$i readers"
+        s"TRef(state=$currentState, lock=$lockState)"
+    end toString
 end TRefImpl
 
 object TRef:

--- a/kyo-stm/shared/src/test/scala/kyo/STMTest.scala
+++ b/kyo-stm/shared/src/test/scala/kyo/STMTest.scala
@@ -84,6 +84,63 @@ class STMTest extends Test:
                 count <- counter.get
             yield assert(result.isFailure && count == 4)
         }
+
+        "arbitrary failure is retried if the transaction is inconsistent" - {
+
+            "within retry budget" in run {
+                for
+                    ref      <- TRef.init(0)
+                    latch1   <- Latch.init(1)
+                    latch2   <- Latch.init(1)
+                    attempts <- AtomicInt.init
+                    _ <-
+                        Async.run {
+                            STM.run(latch1.release.andThen(ref.set(42)))
+                                .andThen(latch2.release)
+                        }
+                    v <-
+                        STM.run {
+                            for
+                                _ <- attempts.incrementAndGet
+                                _ <- latch1.await
+                                _ <- ref.get
+                                _ <- latch2.await
+                                v <- ref.get
+                                _ <- Abort.when(v == 0)(new Exception)
+                            yield v
+                        }
+                    a <- attempts.get
+                yield assert(v == 42 && a == 2)
+            }
+        }
+
+        "exceeding retry budget" in run {
+            for
+                ref      <- TRef.init(0)
+                latch1   <- Latch.init(1)
+                latch2   <- Latch.init(1)
+                attempts <- AtomicInt.init
+                _ <-
+                    Async.run {
+                        STM.run(latch1.release.andThen(ref.set(42)))
+                            .andThen(latch2.release)
+                    }
+                v <-
+                    Abort.run {
+                        STM.run(Schedule.never) {
+                            for
+                                _ <- attempts.incrementAndGet
+                                _ <- latch1.await
+                                _ <- ref.get
+                                _ <- latch2.await
+                                v <- ref.get
+                                _ <- Abort.when(v == 0)(new Exception)
+                            yield v
+                        }
+                    }
+                a <- attempts.get
+            yield assert(v.isFailure && a == 1)
+        }
     }
 
     "with isolates" - {

